File: spec/requests/api/v1/merchants_spec.rb
require "rails_helper"

RSpec.describe "Merchants", type: :request do
  let(:user) { create(:user) }
  let(:headers) { auth_headers(user) }

  it "returns at most ten matching merchants ordered by usage" do
    11.times { |index| create(:merchant, user: user, name: "Coffee #{index}", usage_count: index) }
    create(:merchant, user: user, name: "Tea", usage_count: 100)
    create(:merchant, name: "Coffee outsider", usage_count: 200)

    get "/api/v1/merchants", params: { q: "coffee" }, headers: headers

    expect(response).to have_http_status(:ok)
    expect(json.fetch("data").length).to eq(10)
    expect(json.fetch("data").first["name"]).to eq("Coffee 10")
    expect(json.fetch("data").pluck("name")).not_to include("Coffee outsider", "Coffee 0")
  end

  it "returns every merchant when no search query is given" do
    12.times { |index| create(:merchant, user: user, name: "Merchant #{index}", usage_count: index) }
    create(:merchant, name: "Outsider")

    get "/api/v1/merchants", headers: headers

    expect(response).to have_http_status(:ok)
    expect(json.fetch("data").length).to eq(12)
    expect(json.fetch("data").pluck("name")).not_to include("Outsider")
  end

  it "creates a merchant and nullifies its default category when the category is deleted" do
    category = user.categories.expense.first
    post "/api/v1/merchants", params: { name: "Cafe", default_category_id: category.id }, headers: headers, as: :json
    expect(response).to have_http_status(:created)
    merchant_id = json.dig("data", "id")

    delete "/api/v1/categories/#{category.id}", headers: headers

    expect(response).to have_http_status(:no_content)
    expect(Merchant.find(merchant_id).default_category_id).to be_nil
  end

  it "does not allow another user's category as the default" do
    category = create(:category)

    post "/api/v1/merchants", params: { name: "Cafe", default_category_id: category.id }, headers: headers, as: :json

    expect(response).to have_http_status(:unprocessable_content)
  end

  it "updates a merchant's name and default category" do
    merchant = create(:merchant, user: user, name: "Cafe")
    category = user.categories.expense.first

    patch "/api/v1/merchants/#{merchant.id}", params: { name: "Coffee Shop", default_category_id: category.id }, headers: headers, as: :json

    expect(response).to have_http_status(:ok)
    expect(json.dig("data", "name")).to eq("Coffee Shop")
    expect(json.dig("data", "default_category_id")).to eq(category.id)
    expect(merchant.reload.default_category_id).to eq(category.id)
  end

  it "rejects an update that points at another user's category" do
    merchant = create(:merchant, user: user)
    category = create(:category)

    patch "/api/v1/merchants/#{merchant.id}", params: { default_category_id: category.id }, headers: headers, as: :json

    expect(response).to have_http_status(:unprocessable_content)
    expect(merchant.reload.default_category_id).to be_nil
  end

  it "does not update another user's merchant" do
    merchant = create(:merchant)

    patch "/api/v1/merchants/#{merchant.id}", params: { name: "Hacked" }, headers: headers, as: :json

    expect(response).to have_http_status(:not_found)
    expect(merchant.reload.name).not_to eq("Hacked")
  end
end

File: test/merchants.spec.ts
import {INestApplication} from "@nestjs/common";
import {afterAll, beforeAll, beforeEach, describe, expect, it} from "vitest";

import {DataSource} from "typeorm";
import {createApp, http, HttpClient, resetDatabase} from "./helpers/app";
import {authHeader, createCategory, createMerchant, registerAndLogin, registerNamed} from "./helpers/fixtures";

describe("merchants", () => {
    let app: INestApplication;
    let client: HttpClient;
    let dataSource: DataSource;

    beforeAll(async () => {
        app = await createApp();
        client = http(app);
        dataSource = app.get(DataSource);
    });

    beforeEach(async () => {
        await resetDatabase(dataSource);
    });

    afterAll(async () => {
        await app.close();
    });

    it("search returns at most ten ordered by usage", async () => {
        const fixture = await registerAndLogin(client);
        for (let index = 0; index < 11; index += 1) {
            await createMerchant(dataSource, fixture.userId, `Coffee ${index}`, index, null);
        }
        await createMerchant(dataSource, fixture.userId, "Tea", 100, null);
        const bob = await registerNamed(client, "bob");
        await createMerchant(dataSource, bob.userId, "Coffee outsider", 200, null);

        const response = await client.get("/api/v1/merchants?q=coffee").set(authHeader(fixture.token));
        expect(response.status).toBe(200);
        const rows = response.body.data as Array<Record<string, unknown>>;
        expect(rows).toHaveLength(10);
        expect(rows[0].name).toBe("Coffee 10");
        expect(rows.some(row => row.name === "Coffee outsider")).toBe(false);
        expect(rows.some(row => row.name === "Coffee 0")).toBe(false);
    });

    it("no query returns all owned merchants", async () => {
        const fixture = await registerAndLogin(client);
        for (let index = 0; index < 12; index += 1) {
            await createMerchant(dataSource, fixture.userId, `Merchant ${index}`, index, null);
        }
        const bob = await registerNamed(client, "bob");
        await createMerchant(dataSource, bob.userId, "Outsider", 0, null);

        const response = await client.get("/api/v1/merchants").set(authHeader(fixture.token));
        const rows = response.body.data as Array<Record<string, unknown>>;
        expect(rows).toHaveLength(12);
        expect(rows.some(row => row.name === "Outsider")).toBe(false);
    });

    it("deleting a category nullifies a merchant default category", async () => {
        const fixture = await registerAndLogin(client);
        const headers = authHeader(fixture.token);

        const created = await client.post("/api/v1/merchants").set(headers).send({name: "Cafe", default_category_id: fixture.expenseCategoryId});
        expect(created.status).toBe(201);
        const merchantId = created.body.data.id as string;

        const deleted = await client.delete(`/api/v1/categories/${fixture.expenseCategoryId}`).set(headers);
        expect(deleted.status).toBe(204);

        const list = await client.get("/api/v1/merchants?q=Cafe").set(headers);
        const merchant = (list.body.data as Array<Record<string, unknown>>).find(row => row.id === merchantId);
        expect(merchant?.default_category_id).toBeNull();
    });

    it("rejects another users category as default", async () => {
        const fixture = await registerAndLogin(client);
        const bob = await registerNamed(client, "bob");
        const bobCategory = await createCategory(dataSource, bob.userId, "Bob 分類", 1);

        const response = await client.post("/api/v1/merchants").set(authHeader(fixture.token)).send({name: "Cafe", default_category_id: bobCategory.id});
        expect(response.status).toBe(422);
    });

    it("updates a merchants name and default category", async () => {
        const fixture = await registerAndLogin(client);
        const merchant = await createMerchant(dataSource, fixture.userId, "Cafe", 0, null);

        const response = await client.patch(`/api/v1/merchants/${merchant.id}`).set(authHeader(fixture.token)).send({name: "Coffee Shop", default_category_id: fixture.expenseCategoryId});
        expect(response.status).toBe(200);
        expect(response.body.data.name).toBe("Coffee Shop");
        expect(response.body.data.default_category_id).toBe(fixture.expenseCategoryId);
    });

    it("rejects updating to another users category", async () => {
        const fixture = await registerAndLogin(client);
        const merchant = await createMerchant(dataSource, fixture.userId, "Cafe", 0, null);
        const bob = await registerNamed(client, "bob");
        const bobCategory = await createCategory(dataSource, bob.userId, "Bob 分類", 1);

        const response = await client.patch(`/api/v1/merchants/${merchant.id}`).set(authHeader(fixture.token)).send({default_category_id: bobCategory.id});
        expect(response.status).toBe(422);
    });

    it("does not update another users merchant", async () => {
        const fixture = await registerAndLogin(client);
        const bob = await registerNamed(client, "bob");
        const bobMerchant = await createMerchant(dataSource, bob.userId, "Cafe", 0, null);

        const response = await client.patch(`/api/v1/merchants/${bobMerchant.id}`).set(authHeader(fixture.token)).send({name: "Hacked"});
        expect(response.status).toBe(404);
    });
});
